program-test/src/programs.rs

```diff
@@ -21,76 +26,198 @@ mod spl_associated_token_account {
-    ),
+#[derive(Error, Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ProgramLoadError {
+    #[error("Unknown program ID: {0}")]
+    UnknownId(Pubkey),
+}
+
+struct ProgramInfo {
+    program_id: Pubkey,
+    loader_id: Pubkey,
+    elf: &'static [u8],
+}
+
+static SPL_PROGRAMS: &[ProgramInfo] = &[
+    ProgramInfo {
+        program_id: spl_token::ID,
+        loader_id: bpf_loader::ID,
+        elf: include_bytes!("programs/spl_token-3.5.0.so"),
+    },
+    ProgramInfo {
+        program_id: spl_token_2022::ID,
+        loader_id: bpf_loader_upgradeable::ID,
+        elf: include_bytes!("programs/spl_token_2022-1.0.0.so"),
+    },
+    ProgramInfo {
+        program_id: spl_memo_1_0::ID,
+        loader_id: bpf_loader::ID,
+        elf: include_bytes!("programs/spl_memo-1.0.0.so"),
+    },
+    ProgramInfo {
+        program_id: spl_memo_3_0::ID,
+        loader_id: bpf_loader::ID,
+        elf: include_bytes!("programs/spl_memo-3.0.0.so"),
+    },
+    ProgramInfo {
+        program_id: spl_associated_token_account::ID,
+        loader_id: bpf_loader::ID,
+        elf: include_bytes!("programs/spl_associated_token_account-1.1.1.so"),
+    },
 ];
 
+/// Describes one or more accounts that belong to the same program.
+pub enum ProgramAccounts {
+    /// Non-upgradable programs, one that use the [`bpf_loader`] loader, use a single account.  This
+    /// account will just directly hold ELF of the program.
+    NonUpgradable {
+        id: Pubkey,
+        account: AccountSharedData,
+    },
+
+    /// Upgradable programs, ones that use the [`bpf_loader_upgradeable`] loader, use two accounts.
+    ///
+    /// ELF data is stored in the `program_data_account`, but it is prefixed by an additional
+    /// header.  See [`UpgradeableLoaderState::ProgramData`].
+    Upgradable {
+        main_id: Pubkey,
+        main_account: AccountSharedData,
+        program_data_id: Pubkey,
+        program_data_account: AccountSharedData,
+    },
+}
+
+impl From<ProgramAccounts> for Vec<(Pubkey, AccountSharedData)> {
+    fn from(accounts: ProgramAccounts) -> Self {
+        match accounts {
+            ProgramAccounts::NonUpgradable { id, account } => vec![(id, account)],
+            ProgramAccounts::Upgradable {
+                main_id,
+                main_account,
+                program_data_id,
+                program_data_account,
+            } => vec![
+                (main_id, main_account),
+                (program_data_id, program_data_account),
+            ],
+        }
+    }
+}
+
+/// Returns one or more accounts for the specified program, and the program ELF.
+///
+/// For non-upgradable programs, `elf` data will be the same as the one stored in the program
+/// account.  But for the upgradable programs the relationship is a bit more complex and it could be
+/// convenient to retrieve the ELF bytes directly.
+pub fn spl_program(
+    rent: &Rent,
+    target_id: Pubkey,
+) -> Result<(Vec<u8>, ProgramAccounts), ProgramLoadError> {
```

Comment:
What do you think about flipping the tuple around and calling this `get_spl_program_with_elf_bytes`?

Reply:
Personally I'm on the side of believing it always makes more sense for the function signature to describe what's going to be returned, especially if it's returning a tuple, where only one element is strictly typed (via struct).

Without mentioning `elf` anywhere, this function appears to return some bytes and a `ProgramAccounts` type. One might infer these bytes are ELF bytes, but it's not explicitly known unless they view the function's source or documentation.

With that being said, I guess the documentation provides enough information for this to be used as-is, if you insist on the simple name!

> What is your reasoning for putting them in the opposite order?

Mostly for the naming convention of `spl_program_with_*`, nothing more.

Reply:
Sorry for swooping in here, but please generally avoid returning opaque tuples (especially the one that has 3 components on L200 😬 )

Reply:
Nice! With the refactoring and the type alias this is much easier to follow.